nomos-core/chain-defs/src/tx/bundle.rs

```diff
@@ -0,0 +1,151 @@
+// std
+//crates
+use bytes::{Bytes, BytesMut};
+use risc0_zkvm::Prover;
+use serde::{de::Error as _, Deserialize, Deserializer, Serialize, Serializer};
+//internal
+use super::Error;
+use crate::wire;
+use nomos_proof_statements::bundle::{BundlePrivate, BundlePublic};
+
+#[derive(Debug, Clone)]
+pub struct Bundle {
+    bundle: cl::Bundle,
+    //TODO: this can be pruned once validated
+    proof: risc0_zkvm::Receipt,
+}
+
+impl Bundle {
+    pub fn bundle(&self) -> &cl::Bundle {
+        &self.bundle
+    }
+
+    pub fn prove(bundle_witness: &cl::BundleWitness, prover: &dyn Prover) -> Result<Self, Error> {
+        // need to show that bundle is balanced.
+        // i.e. the sum of ptx balances is 0
+        let bundle_private = BundlePrivate {
+            balances: bundle_witness
+                .partial_witnesses()
+                .iter()
+                .map(|ptx| ptx.balance())
+                .collect(),
+        };
+
+        let env = risc0_zkvm::ExecutorEnv::builder()
+            .write(&bundle_private)
+            .unwrap()
+            .build()
+            .unwrap();
+
+        let start_t = std::time::Instant::now();
+
+        let opts = risc0_zkvm::ProverOpts::groth16();
+        let prove_info = prover
+            .prove_with_opts(env, nomos_cl_risc0_proofs::BUNDLE_ELF, &opts)
+            .map_err(Error::Risc0ProofFailed)?;
+
+        tracing::trace!(
+            "STARK 'bundle' prover time: {:.2?}, total_cycles: {}",
+            start_t.elapsed(),
+            prove_info.stats.total_cycles
+        );
+
+        let receipt = prove_info.receipt;
+
+        Ok(Self {
+            bundle: bundle_witness.commit(),
+            proof: receipt,
+        })
+    }
+
+    pub fn as_bytes(&self) -> Bytes {
+        let mut buff = BytesMut::new();
+        wire::serializer_into_buffer(&mut buff)
+            .serialize_into(&self)
+            .expect("MockTransaction serialization to buffer failed");
+        buff.freeze()
+    }
+}
+
+impl Serialize for Bundle {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        (&self.bundle, &self.proof).serialize(serializer)
```

Comment:
Because I want to do validation on deserialization so I need to write deserialization myself. I also want to write the serialization so that I'm sure they match. A tuple is probably not the best way but it's the most convenient to write. We can change the format later